text/0000-pin_and_move.md

```diff
@@ -0,0 +1,288 @@
+[alternatives]: #alternatives
+
+## Comparison to `?Move`
+
+One previous proposal was to add a built-in `Move` trait, similar to `Sized`. A
+type that did not implement `Move` could not be moved after it had been
+referenced.
+
+This solution had some problems. First, the `?Move` bound ended up "infecting"
+many different APIs where it wasn't relevant, and introduced a breaking change
+in several cases where the API bound changed in a non-backwards compatible way.
+
+In a certain sense, this proposal is a much more narrowly scoped version of
+`?Move`. With `?Move`, *any* reference could act as the "Pin" reference does
+here. However, because of this flexibility, the negative consequences of having
+a type that can't be moved had a much broader impact.
+
+Instead, we require APIs to opt into supporting immovability (a niche case) by
+operating with the `Pin` type, avoiding "infecting" the basic reference type
+with concerns around immovable types.
+
+## Comparison to using `unsafe` APIs
+
+Another alternative we've considered was to just have the APIs which require
+immovability be `unsafe`. It would be up to the users of these APIs to review
+and guarantee that they never moved the self-referential types. For example,
+generator would look like this:
+
+```rust
+trait Generator {
+    type Yield;
+    type Return;
+
+    unsafe fn resume(&mut self) -> CoResult<Self::Yield, Self::Return>;
+}
+```
+
+This would require no extensions to the standard library, but would place the
+burden on every user who wants to call resume to guarantee (at the risk of
+memory insafety) that their types were not moved, or that they were moveable.
+This seemed like a worse trade off than adding these APIs.
+
+## Anchor as a wrapper type and `StableDeref`
+
+In a previous iteration of this RFC, `Anchor` was a wrapper type that could
+"anchor" any smart pointer, and there was a hierarchy of traits relating to the
+stability of the referent of different pointer types.
+
+The primary benefit of this approach was that it was partially integrated with
+crates like owning-ref and rental, which also use a hierarchy of stability
+traits. However, because of differences in the requirements, the traits used by
+owning-ref et al ended up being a non-overlapping subset of the traits proposed
+by this RFC from the traits used by the Anchor type. Merging these into a
+single hierarchy provided relatively little benefit.
+
+And the only types that implemented all of the necessary traits to be put into
+an Anchor before were `Box<T>` and `Vec<T>`. Because you cannot get mutable
+access to the smart pointer (unless the referent implements `Move`), an
+`Anchor<Vec<T>>` was not really any different from an `Anchor<Box<[T]>>` in the
+previous iteration of the RFC. For this reason, making `Anchor` always a box,
+and just supporting `Anchor<[T]>`, reduced the API complexity without losing
+any expressiveness.
+
+## Stack pinning API (potential future extension)
```

Comment:
The stack pinning API @withoutboats proposed in his blog seems to be unsound:

https://play.rust-lang.org/?gist=deb04172b092f72ecf860d95a3b93c1f&version=stable

Reply:
That's not the complete API. The type in that gist named `Pin` is an intermediate step, in this RFC would be called something like `StackPin`. Names and exact shape have shifted around a lot. You then have to apply this function:

```rust
fn(pin: &'a StackPin<'a, T>) -> Pin<'a, T> {
    Pin { data: &pin.data }
}
```

Once you do, the data cannot move.

Reply:
I see